test/GoodGhosting_Attacker_Sends_DAI_Externally.test.js

```diff
@@ -0,0 +1,166 @@
+
+                // Player 1 (index 0 - loser), performs an early withdraw on first segment.
+                if (segmentIndex === 1) {
+                    const earlyWithdrawResult = await goodGhosting.earlyWithdraw({ from: loser});
+                    truffleAssert.eventEmitted(
+                        earlyWithdrawResult,
+                        "EarlyWithdrawal",
+                        (ev) => ev.player === loser,
+                        "loser unable to early withdraw from game",
+                    );
+                }
+
+                // j must start at 1 - Player1 (index 0) early withdraw, so won't continue making deposits
+                for (let j = 1; j < players.length; j++) {
+                    const player = players[j];
+                    const depositResult = await goodGhosting.makeDeposit({ from: player });
+                    truffleAssert.eventEmitted(
+                        depositResult,
+                        "Deposit",
+                        (ev) => ev.player === player && ev.segment.toNumber() === segmentIndex,
+                        `player ${j} unable to deposit for segment ${segmentIndex}`,
+                    );
+                }
+            }
+            // accounted for 1st deposit window
+            // the loop will run till segmentCount - 1
+            // after that funds for the last segment are deposited to protocol then we wait for segment length to deposit to the protocol
+            // and another segment where the last segment deposit can generate yield
+            await timeMachine.advanceTime(segmentLength);
+            await goodGhosting.depositIntoExternalPool({ from: admin });
+            await timeMachine.advanceTime(segmentLength);
+        });
+
+
+        it("redeems funds from external pool", async () => {
+            let eventAmount = new BN(0);
+            const result = await goodGhosting.redeemFromExternalPool({ from: admin });
+            const contractsDaiBalance = new BN(await token.methods.balanceOf(goodGhosting.address).call({ from: admin }));
+            console.log("contractsDaiBalance", contractsDaiBalance.toString());
+            truffleAssert.eventEmitted(
+                result,
+                "FundsRedeemedFromExternalPool",
+                (ev) => {
+                    console.log("totalContractAmount", ev.totalAmount.toString());
+                    console.log("totalGamePrincipal", ev.totalGamePrincipal.toString());
+                    console.log("totalGameInterest", ev.totalGameInterest.toString());
+                    console.log("interestPerPlayer", ev.totalGameInterest.div(new BN(players.length - 1)).toString());
+                    eventAmount = new BN(ev.totalAmount.toString());
+                    return eventAmount.eq(contractsDaiBalance);
+                },
+                `FundsRedeemedFromExternalPool error - event amount: ${eventAmount.toString()}; expectAmount: ${contractsDaiBalance.toString()}`,
```

Comment:
I'd like to add some validation that checks and makes sure that extra funds are correctly distributed to the winners (and only winners). What do you think?

Reply:
will add